File: ldd-kit/adapters/nodejs/telemetry/logging.ts
/**
 * Structured logging module using `pino` — the fastest Node.js JSON logger.
 *
 * Provides JSON/console log rendering, async context propagation via
 * AsyncLocalStorage, and a withLogContext helper for binding trace/tenant/user
 * identifiers.
 *
 * @module telemetry/logging
 */

import pino, { Logger as PinoLogger } from 'pino';
import { AsyncLocalStorage } from 'async_hooks';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Supported log formats. */
export type LogFormat = 'json' | 'console';

/** Supported log levels. */
export type LogLevel = 'debug' | 'info' | 'warn' | 'error' | 'fatal' | 'trace';

/** Logger instance exported from this module. */
export type Logger = PinoLogger;

// ---------------------------------------------------------------------------
// Internal state
// ---------------------------------------------------------------------------

/** Global root logger instance. */
let rootLogger: PinoLogger = pino({ level: 'info' });

/** Per-name logger cache. */
const namedLoggers = new Map<string, PinoLogger>();

/** AsyncLocalStorage for log context propagation across async boundaries. */
const logContextStorage = new AsyncLocalStorage<Record<string, unknown>>();

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

/**
 * Configure the global pino logger with the specified format and level.
 *
 * @param format - `"json"` for production (machine-readable), `"console"` for
 *   development (human-readable with colors).
 * @param level - Log level threshold. One of: `debug`, `info`, `warn`, `error`,
 *   `fatal`, `trace`.
 *
 * @example
 * ```ts
 * configureLogging('json', 'info');
 * ```
 */
export function configureLogging(format: LogFormat, level: LogLevel): void {
  const options: pino.LoggerOptions = {
    level,
    base: { pid: process.pid, hostname: require('os').hostname() },
  };

  if (format === 'console') {
    options.transport = {
      target: 'pino-pretty',
      options: {
        colorize: true,
        translateTime: 'SYS:standard',
        ignore: 'pid,hostname',
      },
    };
  }

  rootLogger = pino(options);

  // Clear named logger cache so new fetches pick up the config.
  namedLoggers.clear();
}

// ---------------------------------------------------------------------------
// Logger factory
// ---------------------------------------------------------------------------

/**
 * Return a named pino Logger. The name is added as the `component` field on
 * every log line.
 *
 * @param name - Logger / module name (e.g. `__filename`).
 * @returns A pino Logger bound with the component name.
 *
 * @example
 * ```ts
 * const logger = getLogger('scan-service');
 * logger.info({ scanId: '123' }, 'scan started');
 * ```
 */
export function getLogger(name: string): PinoLogger {
  const cached = namedLoggers.get(name);
  if (cached) return cached;

  const logger = rootLogger.child({ component: name });
  namedLoggers.set(name, logger);
  return logger;
}

// ---------------------------------------------------------------------------
// Async context propagation
// ---------------------------------------------------------------------------

/**
 * Get the current log context from AsyncLocalStorage.
 * Returns an empty object if no context is active.
 */
function getCurrentContext(): Record<string, unknown> {
  return logContextStorage.getStore() ?? {};
}

/**
 * Execute `fn` within a new async context that carries the provided log
 * fields. Any loggers created or used inside `fn` will automatically merge
 * these fields into every log entry.
 *
 * @param context - Key/value pairs to bind to the async context.
 * @param fn - Function to execute within the context.
 * @returns The return value of `fn`.
 *
 * @example
 * ```ts
 * const result = await withLogContext(
 *   { trace_id: 'abc', tenant_id: 't1' },
 *   async () => {
 *     getLogger('worker').info('processing'); // includes trace_id, tenant_id
 *     return await doWork();
 *   }
 * );
 * ```
 */
export function withLogContext<T>(
  context: Record<string, unknown>,
  fn: () => T
): T {
  const parent = getCurrentContext();
  const merged = { ...parent, ...context };
  return logContextStorage.run(merged, fn);
}

/**
 * Wrap an async function so that it always executes within the current log
 * context, preserving context across promise boundaries.
 *
 * @param fn - Async function to wrap.
 * @returns A new function that preserves log context.
 */
export function withLogContextAsync<TArgs extends unknown[], TReturn>(
  fn: (...args: TArgs) => Promise<TReturn>
): (...args: TArgs) => Promise<TReturn> {
  const snapshot = getCurrentContext();
  return async (...args: TArgs): Promise<TReturn> => {
    return logContextStorage.run(snapshot, () => fn(...args));
  };
}

// ---------------------------------------------------------------------------
// Context-aware logger
// ---------------------------------------------------------------------------

/**
 * Return a logger that automatically includes all fields from the current
 * async log context. This is the preferred way to get a logger inside request
 * handlers or background jobs.
 *
 * @param name - Logger / module name.
 * @returns A pino Logger with async context fields merged in.
 */
export function getContextualLogger(name: string): PinoLogger {
  const base = getLogger(name);
  const context = getCurrentContext();

  if (Object.keys(context).length === 0) {
    return base;
  }

  return base.child(context);
}

// ---------------------------------------------------------------------------
// Convenience exports
// ---------------------------------------------------------------------------

/** The global root logger for direct use. */
export { rootLogger };

/**
 * Log an HTTP request in a consistent structured format.
 *
 * @param logger - pino Logger instance.
 * @param method - HTTP verb.
 * @param path - Request path.
 * @param status - HTTP status code.
 * @param durationMs - Request duration in milliseconds.
 * @param requestId - Unique request identifier.
 */
export function logRequest(
  logger: PinoLogger,
  method: string,
  path: string,
  status: number,
  durationMs: number,
  requestId: string
): void {
  logger.info(
    { method, path, status, duration_ms: durationMs, request_id: requestId },
    'request_completed'
  );
}

/**
 * Log an error in a consistent structured format.
 *
 * @param logger - pino Logger instance.
 * @param err - Error object.
 * @param msg - Human-readable message.
 */
export function logError(logger: PinoLogger, err: unknown, msg: string): void {
  const errorType = err instanceof Error ? err.constructor.name : typeof err;
  const errorMessage = err instanceof Error ? err.message : String(err);

  logger.error(
    { error_type: errorType, error_message: errorMessage, err },
    msg
  );
}

File: ldd-kit/adapters/rust/telemetry/src/logging.rs
//! Structured logging configuration using `tracing` + `tracing-subscriber`.
//!
//! Provides JSON or pretty (console) log formatting, context propagation via
//! `tracing::Span`, and field-based structured logging compatible with the
//! ecosystem of tracing-compatible tools.
//!
//! # Example
//!
//! ```no_run
//! use telemetry::init_logging;
//! use tracing::{info, info_span};
//!
//! init_logging("json", "info");
//!
//! let span = info_span!("process_request", trace_id = "abc123", tenant_id = "t1");
//! let _guard = span.enter();
//! info!(scan_id = "scan-42", "processing roof scan");
//! ```

use tracing_subscriber::{
    fmt::{self, format::FmtSpan},
    layer::SubscriberExt,
    util::SubscriberInitExt,
    EnvFilter,
};

/// Initialise the global `tracing` subscriber with the specified format and
/// level.
///
/// # Arguments
///
/// * `format` — `"json"` for production (machine-readable JSON output) or
///   `"console"` for development (human-readable coloured output).
/// * `level` — Log level threshold. One of: `trace`, `debug`, `info`, `warn`,
///   `error`.
///
/// # Example
///
/// ```no_run
/// use telemetry::init_logging;
///
/// // Production: JSON logs
/// init_logging("json", "info");
///
/// // Development: coloured console output
/// // init_logging("console", "debug");
/// ```
pub fn init_logging(format: &str, level: &str) {
    let env_filter = EnvFilter::try_new(level).unwrap_or_else(|_| EnvFilter::new("info"));

    match format.to_lowercase().as_str() {
        "console" | "pretty" => {
            let fmt_layer = fmt::layer()
                .with_target(true)
                .with_thread_ids(true)
                .with_thread_names(true)
                .with_line_number(true)
                .with_file(true)
                .with_span_events(FmtSpan::CLOSE)
                .pretty();

            tracing_subscriber::registry()
                .with(env_filter)
                .with(fmt_layer)
                .init();
        }
        "json" | _ => {
            let fmt_layer = fmt::layer()
                .json()
                .with_target(true)
                .with_thread_ids(true)
                .with_thread_names(true)
                .with_line_number(true)
                .with_file(true)
                .with_span_events(FmtSpan::CLOSE)
                .with_current_span(true)
                .with_span_list(false);

            tracing_subscriber::registry()
                .with(env_filter)
                .with(fmt_layer)
                .init();
        }
    }
}

/// Initialise logging from environment variables.
///
/// Reads `RUST_LOG` for the log level and `LOG_FORMAT` for the format
/// (`json` or `console`). Falls back to the provided defaults.
///
/// # Arguments
///
/// * `default_format` — Default format if `LOG_FORMAT` is not set.
/// * `default_level` — Default level if `RUST_LOG` is not set.
///
/// # Example
///
/// ```no_run
/// use telemetry::init_logging_from_env;
///
/// init_logging_from_env("json", "info");
/// ```
pub fn init_logging_from_env(default_format: &str, default_level: &str) {
    let format = std::env::var("LOG_FORMAT").unwrap_or_else(|_| default_format.to_string());
    let level = std::env::var("RUST_LOG").unwrap_or_else(|_| default_level.to_string());
    init_logging(&format, &level);
}

// Convenience re-exports of tracing macros — users can use `tracing`
// directly, but these are available for consistency.
pub use tracing::{debug, error, info, trace, warn};
